File: src/main/java/com/educative/grokking/exercises/LongestConsecutiveSequence.java
package com.educative.grokking.exercises;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

// https://leetcode.com/problems/longest-consecutive-sequence/
public class LongestConsecutiveSequence {

    private static int findParent(int num, Map<Integer, Integer> parents) {
        parents.putIfAbsent(num, num);
        if (parents.get(num).equals(num)) {
            return num;
        } else {
            var res = findParent(parents.get(num), parents);
            parents.put(num, res);
            return res;
        }
    }

    private static void union(int a, int b, Map<Integer, Integer> parents, Map<Integer, Integer> ranks, int[] max) {
        var parA = findParent(a, parents);
        var parB = findParent(b, parents);
        ranks.putIfAbsent(parA, 1);
        ranks.putIfAbsent(parB, 1);

        if (parA == parB) {
            return;
        }

        var newRank = ranks.get(parA) + ranks.get(parB);
        if (ranks.get(parA) >= ranks.get(parB)) {
            ranks.put(parA, newRank);
            parents.put(parB, parA);
        } else {
            ranks.put(parB, newRank);
            parents.put(parA, parB);
        }

        if (max[0] < newRank) {
            max[0] = newRank;
        }

    }

    public static int longestConsecutiveSequence(int[] nums) {
        if (nums.length == 0){
            return 0;
        }
        var parents = new HashMap<Integer, Integer>();
        var ranks = new HashMap<Integer, Integer>();
        var numSet = new HashSet<Integer>();
        var maxRank = new int[]{1};
        for (var n : nums) {
            numSet.add(n);
            if (numSet.contains(n - 1)) {
                union(n, n - 1, parents, ranks, maxRank);
            }
            if (numSet.contains(n + 1)) {
                union(n, n + 1, parents, ranks, maxRank);
            }
        }

        return maxRank[0];
    }
}

File: src/main/kotlin/unionfind/LongestConsecutiveSequence.kt
package unionfind

import kotlin.math.max

// https://leetcode.com/problems/longest-consecutive-sequence/description/
class LongestConsecutiveSequence {
    private fun findParent(n: Int, parents: MutableMap<Int, Int>): Int {
        val p1 = parents[n]!!
        if (p1 == n) return n

        val p2 = findParent(p1, parents)
        parents[n] = p2

        return p2
    }

    private fun union(a: Int, b: Int, parents: MutableMap<Int, Int>, ranks: MutableMap<Int, Int>) {
        val parentA = findParent(a, parents)
        val parentB = findParent(b, parents)

        if (parentA == parentB) return

        val aRank = ranks[parentA]!!
        val bRank = ranks[parentB]!!
        if (aRank >= bRank) {
            ranks[parentA] = aRank + bRank
            parents[parentB] = parentA
        } else {
            ranks[parentB] = aRank + bRank
            parents[parentA] = parentB
        }
    }

    fun longestConsecutive(nums: IntArray): Int {
        val parents = nums.map { it to it }.toMap().toMutableMap()
        val ranks = nums.map { it to 1 }.toMap().toMutableMap()

        nums.forEach { n ->
            if (parents.containsKey(n - 1)) {
                union(n, n - 1, parents, ranks)
            }
            if (parents.containsKey(n + 1)) {
                union(n, n + 1, parents, ranks)
            }
        }

        return ranks.values.max()
    }

    fun longestConsecutiveAlternative(nums: IntArray): Int {
        if (nums.isEmpty()) return 0
        if (nums.size == 1) return 1
        val hashSet = HashSet<Int>()
        nums.forEach { hashSet.add(it) }
        var longestSize = 0
        var isNumberStartOfSequence: Boolean
        for (num in nums) {
            isNumberStartOfSequence = !hashSet.contains(num - 1)
            if (isNumberStartOfSequence) {
                var nextConsecutiveNumber = num + 1
                var currentSize = 1
                while (hashSet.contains(nextConsecutiveNumber)) {
                    nextConsecutiveNumber++
                    currentSize++
                }
                longestSize = max(longestSize, currentSize)
            }
        }
        return longestSize
    }
}
